### PR TITLE
feat: domain wide delegation UI

### PR DESCRIPTION
**Pull request by Udit-takkar**
## What does this PR do?

- Fixes #XXXX (GitHub issue number)
- Fixes CAL-XXXX (Linear issue number - should be visible at the bottom of the GitHub issue description)

First PR for domain wide delegation

- UI for adding domain wide delegation and workspace platform
- Team Feature flags check

Splitting a large PR https://github.com/calcom/cal.com/pull/16622 into smaller ones

https://github.com/user-attachments/assets/4cd96973-e063-44c9-a4ff-f5d3d171a331

Screenshot 2024-12-05 at 4 32 30 PM

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] N/A I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox.
- [x] N/A I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

1) Make sure you are logged in as admin user and Enable domain wide delegation  in settings/admin/flags 
2) Go to settings/admin/workspace-platforms to add workspace platform and edit workspace platform
3) Go to settings/organizations/domain-wide-delegation to test adding and removing dwd

Also create a team feature entry to enable it

### CODE REVIEW

File: packages/prisma/migrations/20241205095738_add_feature_flag_domain_wide_delegation/migration.sql [hunk 1]
```diff
@@ -0,0 +1,9 @@
+INSERT INTO
+  "Feature" (slug, enabled, description, "type")
+VALUES
+  (
+    'domain-wide-delegation',
+    false,
+    'Enable Google Workspace Domain-Wide Delegation - Allows system to act on behalf of Google Workspace users.',
+    'OPERATIONAL'
+  ) ON CONFLICT (slug) DO NOTHING;
```
Comment (Udit-takkar) [inline packages/prisma/migrations/20241205095738_add_feature_flag_domain_wide_delegation/migration.sql:9]: Feature flag

File: packages/features/flags/hooks/index.ts [hunk 2]
```diff
@@ -14,6 +14,7 @@ const initialData: AppFlags = {
   "disable-signup": false,
   attributes: false,
   "organizer-request-email-v2": false,
+  "domain-wide-delegation": false,
```
Comment (hariombalhara) [inline packages/features/flags/hooks/index.ts:17]: We should make it a teamFeature like Google Calendar cache

**Review by hariombalhara — DISMISSED** [review]
Let's implement a team level feature flag and will test after that !!

File: packages/features/domain-wide-delegation/domain-wide-delegation.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  static async init(userId: number, teamId: number | null) {
+    const domainWideDelegationEnabled = await this.checkIfDwDIsEnabled(userId, teamId);
+
+    if (!domainWideDelegationEnabled) {
+      return new MockDomainWideDelegationRepository();
+    }
+
+    return new DomainWideDelegationRepository();
+  }
+}
```
Comment (Udit-takkar) [inline packages/features/domain-wide-delegation/domain-wide-delegation.ts:37]: For implementing team based feature and only enable Dwd for selected teams

File: packages/features/ee/organizations/pages/settings/domainWideDelegation.tsx [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+
+  if (error) {
+    return <div>{t(error.message)}</div>;
+  }
+
+  if (!delegations || !workspacePlatforms) {
+    return <div>{t("something_went_wrong")}</div>;
+  }
+
+  return (
```
Comment (Udit-takkar) [inline packages/features/ee/organizations/pages/settings/domainWideDelegation.tsx:371]: Just some UI improvements

File: packages/trpc/server/routers/viewer/domainWideDelegation/_router.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
+import authedProcedure, { authedOrgAdminProcedure } from "../../../procedures/authedProcedure";
```
Comment (Udit-takkar) [inline packages/trpc/server/routers/viewer/domainWideDelegation/_router.ts:1]: No changes from the old PR

File: packages/prisma/migrations/20241204193501_add_domain_wide_delegation/migration.sql [hunk 6]
```diff
@@ -0,0 +1,57 @@
+-- AlterTable
```
Comment (Udit-takkar) [inline packages/prisma/migrations/20241204193501_add_domain_wide_delegation/migration.sql:1]: All dwd schema changes here

File: packages/lib/server/repository/workspacePlatform.ts [hunk 7]
```diff
@@ -0,0 +1,93 @@
+import type { Prisma } from "@prisma/client";
```
Comment (Udit-takkar) [inline packages/lib/server/repository/workspacePlatform.ts:1]: No change from old PR

File: apps/web/app/settings/(admin-layout)/admin/workspace-platforms/page.tsx [hunk 8]
```diff
@@ -0,0 +1,28 @@
+import { _generateMetadata } from "app/_utils";
```
Comment (Udit-takkar) [inline apps/web/app/settings/(admin-layout)/admin/workspace-platforms/page.tsx:1]: Moved both of the pages to app dir

File: packages/prisma/schema.prisma [hunk 9]
```diff
@@ -564,8 +568,10 @@ model BookingReference {
   externalCalendarId         String?
   deleted                    Boolean?
 
-  credential   Credential? @relation(fields: [credentialId], references: [id], onDelete: SetNull)
-  credentialId Int?
+  credential    Credential?           @relation(fields: [credentialId], references: [id], onDelete: SetNull)
+  credentialId  Int?
+  delegatedTo   DomainWideDelegation? @relation(fields: [delegatedToId], references: [id], onDelete: Cascade)
```
Comment (hariombalhara) [inline packages/prisma/schema.prisma:573]: I don't see why I called it `delegatedToId` here and `domainWideDelegationCredentialId` in DestinationCalendar and SelectedCalendar. Seems like a miss from my side.

I think it should be quick to change it to `domainWideDelegationCredentialId` and `domainWideDelegationCredential`. Also, make sure to regenerate migrations to have no mention of delegatedTo in there

File: packages/prisma/schema.prisma [hunk 10]
```diff
@@ -1638,6 +1646,46 @@ enum EventTypeAutoTranslatedField {
...
+  organizationId      Int
+  organization        Team                  @relation(fields: [organizationId], references: [id], onDelete: Cascade)
+  domain              String
+  selectedCalendars   SelectedCalendar[]
+  destinationCalendar DestinationCalendar[]
+  bookingReferences   BookingReference[]
+  createdAt           DateTime              @default(now())
+  updatedAt           DateTime              @updatedAt
+
+  // Should be fare to assume that one domain can be only on one workspace platform at a time. So, one can't have two different workspace platforms for the same domain
```
Comment (hariombalhara) [inline packages/prisma/schema.prisma:1666]:
```suggestion
  // Should be fair to assume that one domain can be only on one workspace platform at a time. So, one can't have two different workspace platforms for the same domain
```

File: packages/features/domain-wide-delegation/domain-wide-delegation-repository.interface.ts
Comment (hariombalhara) [inline packages/features/domain-wide-delegation/domain-wide-delegation-repository.interface.ts:1]: What is the purpose of this file? It wasn't there in the original PR.

File: packages/lib/server/repository/domainWideDelegation.ts [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+    },
+  },
+};
+
+const domainWideDelegationSelectIncludesServiceAccountKey = {
+  ...domainWideDelegationSafeSelect,
+  serviceAccountKey: true,
+};
+
+export class DomainWideDelegationRepository implements IDomainWideDelegationRepository {
```
Comment (hariombalhara) [inline packages/lib/server/repository/domainWideDelegation.ts:42]: Why do we need an interface for this?

Comment (Udit-takkar) [reply packages/lib/server/repository/domainWideDelegation.ts:42]: Just to make sure if we add a new method to this repository we also add that method to mock repository

Comment (hariombalhara) [reply packages/lib/server/repository/domainWideDelegation.ts:42]: But why do we need MockRepository?

Comment (Udit-takkar) [reply packages/lib/server/repository/domainWideDelegation.ts:42]: It's just cleaner way to not expose the methods of domainWideDelegation when it is disabled. similar to packages/features/calendar-cache/calendar-cache.ts

Comment (hariombalhara) [reply packages/lib/server/repository/domainWideDelegation.ts:42]: Okay, just went into detail and noticed it is a pattern introduced in Calendar Cache webhook PR. We should be fine for now but I will dig deep if we can avoid maintaining these mock files and interface files later.

File: packages/prisma/schema.prisma (hunk 9 above)
Comment (Udit-takkar) [reply packages/prisma/schema.prisma:573]: > I don't see why I called it delegatedToId here and domainWideDelegationCredentialId in DestinationCalendar and SelectedCalendar. Seems like a miss from my side.

Actually i added this field. let me fix the name

File: packages/features/flags/config.ts [hunk 12]
```diff
@@ -15,4 +15,5 @@ export type AppFlags = {
   "disable-signup": boolean;
   attributes: boolean;
   "organizer-request-email-v2": boolean;
+  "domain-wide-delegation": boolean;
```
Comment (hariombalhara) [inline packages/features/flags/config.ts:18]: We might not need it now.

Comment (Udit-takkar) [reply packages/features/flags/config.ts:18]: team feature flag requires a featureSlug taken from Feature table

Comment (hariombalhara) [reply packages/features/flags/config.ts:18]: Thanks!! makes sense

File: packages/prisma/schema.prisma [hunk 13]
```diff
@@ -564,8 +568,10 @@ model BookingReference {
   externalCalendarId         String?
   deleted                    Boolean?
 
-  credential   Credential? @relation(fields: [credentialId], references: [id], onDelete: SetNull)
-  credentialId Int?
+  credential                       Credential?           @relation(fields: [credentialId], references: [id], onDelete: SetNull)
+  credentialId                     Int?
+  domainWideDelegation             DomainWideDelegation? @relation(fields: [domainWideDelegationCredentialId], references: [id], onDelete: Cascade)
```
Comment (hariombalhara) [inline packages/prisma/schema.prisma:573]: @Udit-takkar Just noticed that I have configured to delete it when DWD is deleted. I think that is wrong. We should configure it to have same behaviour as credential i.e. onDelete:SetNull.

It means that right now if we delete a DomainWideDelegation entry that would delete the BookingReference which could be used in some existing booking.

**Review by hariombalhara — APPROVED** [review]

**Pull request merged**